File: architecture-skills/design-patterns/examples/memento/memento.java
// Memento pattern in Java.
//
// A private nested record gives real opacity: only TextEditor can read
// the snapshot's fields, so History stores state it cannot inspect.

import java.util.ArrayDeque;
import java.util.Deque;

public class memento {

    static final class TextEditor {
        private StringBuilder text = new StringBuilder();
        private int cursor = 0;

        /** Memento: fields visible only to the enclosing originator. */
        record Snapshot(String text, int cursor) {}

        void type(String insert) {
            text.insert(cursor, insert);
            cursor += insert.length();
        }

        Snapshot save() {
            return new Snapshot(text.toString(), cursor);
        }

        void restore(Snapshot s) {
            text = new StringBuilder(s.text());
            cursor = s.cursor();
        }

        @Override
        public String toString() {
            return "'" + text + "' (cursor at " + cursor + ")";
        }
    }

    /** Caretaker: stores snapshots but never inspects them. */
    static final class History {
        private final Deque<TextEditor.Snapshot> snapshots = new ArrayDeque<>();

        void push(TextEditor.Snapshot s) {
            snapshots.push(s);
        }

        TextEditor.Snapshot pop() {
            return snapshots.isEmpty() ? null : snapshots.pop();
        }
    }

    public static void main(String[] args) {
        TextEditor editor = new TextEditor();
        History history = new History();

        editor.type("Hello");
        history.push(editor.save());
        editor.type(", world");
        history.push(editor.save());
        editor.type("!!!");
        System.out.println("current: " + editor);

        for (int i = 0; i < 2; i++) {
            TextEditor.Snapshot s = history.pop();
            if (s != null) {
                editor.restore(s);
                System.out.println("undo:    " + editor);
            }
        }
    }
}

File: architecture-skills/design-patterns/examples/memento/memento.py
"""Memento pattern in Python.

A text editor saves snapshots of its state so a history object can undo
edits without ever looking inside the editor. A frozen dataclass makes
the memento immutable; Python has no enforced privacy, so opacity is by
convention (the caretaker just never touches snapshot fields).
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Snapshot:
    """Memento: immutable, opaque to the caretaker."""

    text: str
    cursor: int


class TextEditor:
    """Originator: the only class that reads or writes snapshot state."""

    def __init__(self) -> None:
        self._text = ""
        self._cursor = 0

    def type(self, text: str) -> None:
        self._text = self._text[: self._cursor] + text + self._text[self._cursor :]
        self._cursor += len(text)

    def save(self) -> Snapshot:
        return Snapshot(self._text, self._cursor)

    def restore(self, snapshot: Snapshot) -> None:
        self._text = snapshot.text
        self._cursor = snapshot.cursor

    def __str__(self) -> str:
        return f"{self._text!r} (cursor at {self._cursor})"


class History:
    """Caretaker: stores mementos but never inspects them."""

    def __init__(self) -> None:
        self._snapshots: list[Snapshot] = []

    def push(self, snapshot: Snapshot) -> None:
        self._snapshots.append(snapshot)

    def pop(self) -> Snapshot | None:
        return self._snapshots.pop() if self._snapshots else None


if __name__ == "__main__":
    editor = TextEditor()
    history = History()

    editor.type("Hello")
    history.push(editor.save())
    editor.type(", world")
    history.push(editor.save())
    editor.type("!!!")
    print("current:", editor)

    for _ in range(2):
        snapshot = history.pop()
        if snapshot:
            editor.restore(snapshot)
            print("undo:   ", editor)
